src/Elastic.Apm/Metrics/MetricsProvider/ProcessTotalCpuTimeProvider.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Api;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	public class ProcessTotalCpuTimeProvider : IMetricsProvider
+	{
+		private const string ProcessCpuTotalPct = "system.process.cpu.total.norm.pct";
+		private readonly object _lock = new object();
+		private bool _first = true;
+		private TimeSpan _lastCurrentProcessCpuTime;
+		private DateTime _lastTick;
+
+		private Version _processAssemblyVersion;
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string NameInLogs => "total process CPU time";
+
+		public IEnumerable<Sample> GetValue()
+		{
+			var timespan = DateTime.UtcNow;
+			var cpuUsage = Process.GetCurrentProcess().TotalProcessorTime;
+			var currentTimeSpan = DateTime.UtcNow;
+
+			if (!_first)
+			{
+				if (_processAssemblyVersion == null)
+					_processAssemblyVersion = typeof(Process).Assembly.GetName().Version;
+
+				double cpuUsedMs;
+
+				//workaround for a CoreFx bug. See: https://github.com/dotnet/corefx/issues/37614#issuecomment-492489373
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && _processAssemblyVersion < new Version(4, 3, 0))
+					cpuUsedMs = (cpuUsage - _lastCurrentProcessCpuTime).TotalMilliseconds / 100;
+				else
+					cpuUsedMs = (cpuUsage - _lastCurrentProcessCpuTime).TotalMilliseconds;
+
+				var totalMsPassed = (currentTimeSpan - _lastTick).TotalMilliseconds;
+				var cpuUsageTotal = cpuUsedMs / (Environment.ProcessorCount * totalMsPassed);
+
+
+				_first = false;
```

Comment:
Not in the first iteration - and this is only called in the first iteration. Without this the code does not work.

Reply:
Are you sure? `_first = false;` statement is inside _then_ branch of `if (!_first)` which means `_first` is `false`. Am I missing something?

Reply:
Yes, you miss the `private bool _first = true;` at the beginning of the method. 

In the 1. iteration `_first` is true, so we are in the _then_ block as you say, set it to false and in the 2 round (and every subsequent round) we execute the `if (!_first)` branch.

Try it with removing that line and run the actual code. You will never get values from this method.

Reply:
Since the code is refactored it's not relevant anymore but if you will run it in debugger you will see that you are wrong: when the method enters for the first when `_first` is `true` (we agree about that) and so you don't enter `if (!_first)` branch. The method jumps over `if` to the line `_first = false; `. Then when the method is called the second time you will enter `if (!_first)` branch which makes sense because it's `if(_first == false)` so obviously inside then branch `_first` is already `false` otherwise then branch wouldn't have been entered.

Reply:
Oh, i mixed this up with another line later below - should have looked at the line number. Anyway, that code is gone.